### DISCUSSION TITLE
Is there any valid reason Rector should run more than once for one file?

### DISCUSSION THREAD

**Discussion by alexkramer98** [rectorphp/rector#7085] [General]
# Question

I came accross this, for example, using the ```AnnotationToAttributeRector``` combined with ```NewlineAfterStatementRector```. The first pass converts the annotations to attributes, and the next pass (re-)adds the newlines. I would expect both of these rules to do their magic in a single pass.

I found this behavior with ECS as well. I understand a bug could be causing the file to change, so that it has to be parsed again for a complete refactoring.

However, I am wondering if there would be any valid reason for this behavior to occur beyond of the scope of bugs -- if there is, how do I know (up front?) how many passes are needed for a file? One, two, maybe even more?

What is your point of view about this?

**Comment by TomasVotruba** [comment]
Hi, it's the way the nodes and tokens are traversed. It usually does not happen, but it depends on use case. Here it's quite clear, as former doblock, not even PHP node, is convertded to a new element, an [`Attribute` node](https://github.com/nikic/PHP-Parser/blob/master/lib/PhpParser/Node/Attribute.php).

E.g. if a rule changes visibliity of method from `private` to `public`, the tool might run again and see the `public` method in the end. It will try to move it up along with other `public` methods.

Nodes are traversed from top to the bottom. See: https://github.com/nikic/PHP-Parser/blob/master/doc/2_Usage_of_basic_components.markdown#node-traversation

Usually, you won't notice these as Rector runs in CI and commits changes untill there is no more.

Still, my knowledge is limited, so maybe there is a way to handle these cases better. If you can see it, please propose a PR so we can save even more work :+1: 

(I've moved this to discussions, as this is not an issue, but rather topic to talk about.)

**Reply by alexkramer98** [reply]
Thanks for your view! That sounds logical.

 We've been triggering Rector on the fly, whenever a file changes. That's why I discovered this behavior.
